applications/ParticleMechanicsApplication/custom_constitutive/johnson_cook_thermal_plastic_3D_law.cpp

```diff
@@ -0,0 +1,476 @@
+
+		mStrainOld = ZeroVector(GetStrainSize());
+		mEquivalentPlasticStrainOld = 0.0;
+		mPlasticStrainRateOld = 0.0;
+		mEnergyInternal = 0.0;
+		mEnergyDissipated = 0.0;
+		mTemperatureOld = rMaterialProperties[TEMPERATURE];
+		mGammaOld = 1e-8;
+		mHardeningRatio = 1.0;
+
+		if (rMaterialProperties[TAYLOR_QUINNEY_COEFFICIENT] == 0.0) {
+			KRATOS_WARNING("Johnson Cook Material Model") << " Taylor Quinney Coefficient set to 0, ignoring thermal effects" << std::endl;
+		}
+
+		mYieldStressOld = CalculateHardenedYieldStress(rMaterialProperties, mEquivalentPlasticStrainOld, mPlasticStrainRateOld, mTemperatureOld);
+		mYieldStressVirgin = mYieldStressOld;
+	}
+
+
+	void JohnsonCookThermalPlastic3DLaw::CalculateMaterialResponseKirchhoff(Kratos::ConstitutiveLaw::Parameters& rValues)
+	{
+		// Check if the constitutive parameters are passed correctly to the law calculation
+		CheckParameters(rValues);
+
+		// Get Values to compute the constitutive law:
+		Flags& Options = rValues.GetOptions();
+		KRATOS_ERROR_IF(Options.Is(ConstitutiveLaw::USE_ELEMENT_PROVIDED_STRAIN))
+			<< "The JohnsonCookThermalPlastic3DLaw cannot accept the deformation graddient F as a strain input."
+			<< " Please set the strain vector and set USE_ELEMENT_PROVIDED_STRAIN = False in the element.";
+		const ProcessInfo& CurrentProcessInfo = rValues.GetProcessInfo();
+		CheckIsExplicitTimeIntegration(CurrentProcessInfo);
+		const Properties& MaterialProperties = rValues.GetMaterialProperties();
+		const double yield_stress_failure_ratio = 1e-3; // particle fails if current_yield/virgin_yield < yield_stress_failure_ratio
+
+		// Get old stress vector and current strain vector
+		const Vector StrainVector = rValues.GetStrainVector();
+		Vector& StressVector = rValues.GetStressVector();
+		const Matrix identity = (GetStrainSize() == 3) ? IdentityMatrix(2) : IdentityMatrix(3);
+
+		// Convert vectors to matrices for easier manipulation
+		Matrix stress_old = (GetStrainSize() == 3) ? Matrix(2, 2) : Matrix(3, 3);
+		Matrix strain_increment = (GetStrainSize() == 3) ? Matrix(2, 2) : Matrix(3, 3);
+		MakeStrainStressMatrixFromVector((StrainVector - mStrainOld), strain_increment);
+		MakeStrainStressMatrixFromVector(StressVector, stress_old);
+
+		// Material moduli
+		const double shear_modulus_G = MaterialProperties[YOUNG_MODULUS] / (2.0 + 2.0 * MaterialProperties[POISSON_RATIO]);
+		const double bulk_modulus_K = MaterialProperties[YOUNG_MODULUS] / (3.0 - 6.0 * MaterialProperties[POISSON_RATIO]);
+
+		// Calculate deviatoric quantities
+		const Matrix strain_increment_deviatoric = strain_increment - CalculateMatrixTrace(strain_increment) / 3.0 * identity;
+		const Matrix stress_deviatoric_old = stress_old - CalculateMatrixTrace(stress_old) / 3.0 * identity;
+
+		// Calculate trial (predicted) j2 stress
+		double stress_hydrostatic_new = CalculateMatrixTrace(stress_old) / 3.0 + bulk_modulus_K * CalculateMatrixTrace(strain_increment);
```

Comment:
Can you check the `mpm_stress_principal_invariants_utility.h` file? This is to reduce duplication of writing stress invariant computation.

Reply:
It's a good idea - I had a look and there is no option just to calculate one invariant by itself (since I only need I1).

Reply:
True. You can add it there I think. This is just to reduce the code duplication, so maybe adding single function will be better there?